### DISCUSSION TITLE
issue: `setValue` cannot be used to set a `Date` input field in typescript without converting the value into a string

### DISCUSSION THREAD

**Discussion by nathanpovo** [react-hook-form/react-hook-form#9318] [Q&A] [answered]
### Version Number

7.39.1

### Codesandbox/Expo snack

https://codesandbox.io/s/tender-sound-8tqzsf?file=/src/App.tsx

[Backup of the `App.tsx` file in the codesandbox](https://gist.github.com/nathanpovo/b28dc3c88cb39e3cadde0a5f327cfd9e)

### Steps to reproduce

1. Open the sandbox

### Expected behaviour

The `setValue` should set the value of the date input from the `Date` object passed to the function.

### Current behaviour

The `setValue` function accepts the `Date` object but does not set the input field's value to it. The `Date` object has to be converted to a string with the format `YYYY-MM-DD` for the value of the date input to be set.

#### Workaround 1

Change the type of the input from just `Date` to `Date | string` and pass in the manually converted `Date` string.

``` typescript
type Inputs = {
  exampleDate: Date | string;
};

// ...

useEffect(() => {
  const date = new Date(2022, 10, 3);

  setValue("exampleDate", ConvertDateToString(date));
}, [setValue]);
```

#### Workaround 2

Wrap the input in a [`Controller` component](https://react-hook-form.com/api/usecontroller/controller) and set the input value using a function that converts the `Date` to a `string` with the correct format.

``` typescript
type Inputs = {
  exampleDate: Date | string;
};

// ...

useEffect(() => {
  const date = new Date(2022, 10, 3);

  setValue("exampleDate", date);
}, [setValue]);

// ...

<Controller
  control={control}
  name="exampleDate"
  render={({ field: { name, onChange, value, ref } }) => (
    <input
      name={name}
      type="date"
      onChange={onChange}
      ref={ref}
      value={ConvertDateToString(value)}
    />
  )}
/>
```

#### Workaround 3

Cast the converted `string` value to `any` before setting the input value.

``` typescript
type Inputs = {
  exampleDate: Date;
};

// ...

useEffect(() => {
  const date = new Date(2022, 10, 3);

  setValue("exampleDate", date as any);
}, [setValue]);
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional Notes

I am not sure if this is the intended behaviour or not. The documentation never mentioned this scenario and I could not find any issues that discussed this problem (other than #8807 which is talking about `defaultValue`).

**Comment by bluebill1049** [comment, marked as answer]
Input `type="date"` accepts a string for defaultValue, hook form doesn't transform your date object into string. Similar to an input `type=number`. whatever you supplied is equivalent to `e.taget.value = ""`

**Reply by nathanpovo** [reply]
Thank you for the (super) quick response.

The issue isn't with `defaultValue`, it is with `setValue`. For a form with a `Date` field, `setValue` will only accept a `Date` object.

In the following typescript react component, you get a compilation error because `setValue` is expecting a `Date` object for its `value` parameter

``` typescript
import { useEffect } from "react";
import { useForm } from "react-hook-form";

interface FormInput {
  myField: Date;
}

const Tester = () => {
  const { register, setValue } = useForm<FormInput>();

  useEffect(() => {
    setValue("myField", "2022-11-03");
  }, [setValue]);

  return (
    <form>
      <input
        type="date"
        {...register("myField", {
          valueAsDate: true,
        })}
      />
    </form>
  );
};

export default Tester;
```

![image](https://user-images.githubusercontent.com/17831160/199726082-4b9a87e2-9d7c-421c-98f1-1dc01f4b6546.png)

However, when passing a `Date` object to the `setValue` function, the date input field is not set to the date that was passed (as you can see in the codesandbox).

![image](https://user-images.githubusercontent.com/17831160/199726683-86b9dad2-daed-4f65-b7da-8345608b64a5.png)

**Reply by Moshyfawn** [reply]
When you're using native Web inputs _(uncontrolled input via `register`)_, the source of truth for your value is the input's value, which can only be a `string`. So, you can
- **either** work with the `string` and format the date value `onSubmit`
- **or** opt-in to use controlled input _(`Controller` / `useController` API)_ and manipulate your date value however you want

**Reply by nathanpovo** [reply]
That makes sense.

The types in the `react-hook-form` package made me think that it would be able to convert the object to a string and the `input` component would be able to handle. I didn't give that much thought to what the `input` component is actually doing under the hood.

Going forward I will use the `Controller` component to set the value of the `input` component.

**Reply by bluebill1049** [reply]
thanks, @Moshyfawn for the detailed explanation. ❤️

**Reply by tan00001** [reply]
If the source of truth for a value is the input's value, we would not need validation anymore now would we? The whole idea of using TypeScript and strong type everything is because we do not want to treat everything as a string. There is already a "valueAsDate" option, yet even with it set, when provided with a Date default value, it gets thrown away. This just makes no sense.

**Comment by leapful** [comment]
Reference to the answer: https://github.com/react-hook-form/react-hook-form/discussions/9318#discussioncomment-4048080
